src/analysis/normed_space/conformal.lean

```diff
@@ -0,0 +1,211 @@
+        ... = c₁ * c₁ * inner u v / (abs c₁ * abs c₁ * ∥u∥ * ∥v∥) : by rw [real.norm_eq_abs]
+        ... = c₁ * c₁ * inner u v / (c₁ * c₁ * ∥u∥ * ∥v∥) : by rw [← pow_two, ← sq_abs, pow_two]
+        ... = c₁ * (c₁ * inner u v) / (c₁ * (c₁ * (∥u∥ * ∥v∥))) : by simp only [mul_assoc]
+        ... = (c₁ * inner u v) / (c₁ * (∥u∥ * ∥v∥)) : by rw mul_div_mul_left _ _ hc₁
+        ... = inner u v / (∥u∥ * ∥v∥) : by rw mul_div_mul_left _ _ hc₁, },
+end
+
+end conformal_at
+
+
+/-- A map `f` is conformal on a set `s` if it's conformal at every point in that set. -/
+def conformal_on {X Y : Type*} [inner_product_space ℝ X] [inner_product_space ℝ Y]
+  (f : X → Y) (s : set X) :=
+∀ (x : X), x ∈ s → conformal_at f x
+
+namespace conformal_on
+
+variables {X Y Z: Type*}
+  [inner_product_space ℝ X] [inner_product_space ℝ Y] [inner_product_space ℝ Z]
+
+theorem conformal_at {f : X → Y} {u : set X} (h : conformal_on f u) {x : X} (hx : x ∈ u) :
+conformal_at f x := h x hx
+
+theorem comp {f : X → Y} {g : Y → Z}
+  {u : set X} {v : set Y} (hf : conformal_on f u) (hg : conformal_on g v) :
+  conformal_on (g ∘ f) (u ∩ f⁻¹' v) :=
+λ x hx, (hf x hx.1).comp (hg (f x) $ set.mem_preimage.mp hx.2)
+
+theorem congr {f : X → X} {g : X → X}
+  {u : set X} (hu : is_open u) (h : ∀ (x : X), x ∈ u → g x = f x) (hf : conformal_on f u) :
+  conformal_on g u := λ x hx, let ⟨f', h₁, c, hc, lie, h₂⟩ := hf x hx in
+begin
+  have : has_fderiv_at g f' x :=
+    by apply h₁.congr_of_eventually_eq; rw filter.eventually_eq_iff_exists_mem;
+       exact ⟨u, hu.mem_nhds hx, h⟩,
+  exact ⟨f', this, c, hc, lie, h₂⟩,
+end
+
+end conformal_on
+
+/-- A map `f` is conformal if it's conformal at every point. -/
+def conformal {X Y : Type*} [inner_product_space ℝ X] [inner_product_space ℝ Y] (f : X → Y) :=
+∀ (x : X), conformal_at f x
+
+namespace conformal
+
+variables {X Y Z : Type*}
+  [inner_product_space ℝ X] [inner_product_space ℝ Y] [inner_product_space ℝ Z]
+
+theorem conformal_at {f : X → Y} (h : conformal f) (x : X) : conformal_at f x := h x
+
+theorem conformal_on {f : X → Y} (h : conformal f) : conformal_on f set.univ := λ x hx, h x
+
+theorem differentiable {f : X → Y} (h : conformal f) : differentiable ℝ f :=
+λ x, (h x).differentiable_at
+
+theorem id : conformal (id : X → X) := λ x, conformal_at.id x
+
+theorem const_smul {c : ℝ} (h : c ≠ 0) : conformal (λ (x : X), c • x) :=
+λ x, conformal_at.const_smul h x
+
+theorem comp {f : X → Y} {g : Y → Z} (hf : conformal f) (hg : conformal g) : conformal (g ∘ f) :=
+λ x, conformal_at.comp (hf x) (hg $ f x)
+
+end conformal
+
+variables {X : Type*} [inner_product_space ℝ X]
+
+/-- The pregroupoid of conformal maps. -/
+def conformal_pregroupoid : pregroupoid X :=
+{ property := λ f u, conformal_on f u,
+  comp := λ f g u v hf hg hu hv huv, hf.comp hg,
+  id_mem := conformal.conformal_on conformal.id,
+  locality := λ f u hu h x hx, let ⟨v, h₁, h₂, h₃⟩ := h x hx in h₃ x ⟨hx, h₂⟩,
+  congr := λ f g u hu h hf, conformal_on.congr hu h hf, }
```

Comment:
Can you please put the manifolds material in a separate file, in the manifolds folder, even if it's very short for now?

Reply:
Thank you! What should I name the new file?

Reply:
Hmm ... maybe either `conformal_groupoid` or `conformally_flat_manifold`?

Reply:
> One other generalization: in the definition `is_conformal_map` can you use `linear_isometry` rather than `linear_isometry_equiv`? At some point the theory will then require a lemma saying that an invertible conformal map (with the weakened definition) is a multiple of a `linear_isometry_equiv`, but I'm not sure the need will turn up in this PR.

Sure. But I used the theorem `linear_isometry_complex` in my proofs about complex conformal maps. Should I add this invertible definition after this not necessarily invertible definition `is_conformal_map`?